src/charm.py

```diff
@@ -138,6 +141,63 @@ def __init__(self, *args):
         self.framework.observe(self.on.config_changed, self._configure)
         self.framework.observe(self.on[PEER_RELATION_NAME].relation_created, self._configure)
         self.framework.observe(self.on[PEER_RELATION_NAME].relation_changed, self._configure)
+        self.framework.observe(self.on.authorize_charm_action, self._on_authorize_charm_action)
+
+    def _on_authorize_charm_action(self, event: ActionEvent):
+        """Authorize the charm to interact with Vault."""
+        if not self.unit.is_leader():
+            event.fail("This action can only be run by the leader unit")
+            return
+        logger.info("Authorizing the charm to interact with Vault")
+        if not self._api_address:
+            event.fail("API address is not available. Something is wrong.")
+            return
+        token = event.params["token"]
+        vault = Vault(
+            url=self._api_address, ca_cert_path=self.tls.get_tls_file_path_in_charm(File.CA)
+        )
+        vault.set_token(token)
+        vault.enable_audit_device(device_type="file", path="stdout")
+        vault.enable_approle_auth()
+        vault.configure_charm_access_policy(name=CHARM_POLICY_NAME)
+        # TODO: What should this actually be set to?
```

Comment:
Are we forced to provide a cidr?

Reply:
Haven't actually tested this yet, I'll look into it.

Reply:
Answer: no.

Is there any concern here if we allow this token to be used by anyone? I'm not familiar enough with Juju's network model to know if we should be concerned about anything here.

Reply:
Well indeed we would like for this token to be only usable by the charm. However if there is no reliable way to ensure this via parameters like cidr's (ex. if we can't count on the charm using the same subnet for its whole lifecycle), let's not use them and focus on having strict policies.